### PR TITLE
PLF-6642:Property exo:lastModifiedDate not found when the daily and w…

### PR DESCRIPTION
**Pull request by mkrout**
…eekly notifications jobs are executed

### CODE REVIEW

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 1]
```diff
@@ -427,24 +426,31 @@ private NodeIterator getDigestIterator(NotificationContext context, SessionProvi
...
-    //
-    model.setLastUpdateTime(node.getParent().getProperty(EXO_LAST_MODIFIED_DATE).getDate());
-    //
-    if (node.hasProperty(EXO_IS_ENABLED)) {
-      model.setEnabled(Boolean.valueOf(node.getProperty(EXO_IS_ENABLED).getString()));
+    if(!node.getParent().hasProperty( EXO_LAST_MODIFIED_DATE)){
+      if(node.getParent().canAddMixin("exo:modify")) {
+        node.getParent().addMixin("exo:modify");
+      }
+      node.getParent().setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
```
Comment (thomasdelhomenie) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: Is it possible to fall in the case of the "exo:modify" has not been added here, so the property cannot be set ?

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 2]
```diff
@@ -427,24 +426,31 @@ private NodeIterator getDigestIterator(NotificationContext context, SessionProvi
...
-    for (AbstractChannel channel : channels) {
-      model.setChannelPlugins(channel.getId(), getValues(node, getChannelProperty(channel.getId())));
-    }
-    //
-    model.setLastUpdateTime(node.getParent().getProperty(EXO_LAST_MODIFIED_DATE).getDate());
-    //
-    if (node.hasProperty(EXO_IS_ENABLED)) {
-      model.setEnabled(Boolean.valueOf(node.getProperty(EXO_IS_ENABLED).getString()));
+    if(!node.getParent().hasProperty( EXO_LAST_MODIFIED_DATE)){
+      if(node.getParent().canAddMixin("exo:modify")) {
```
Comment (thomasdelhomenie) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: "exo:modify" should be extracted as a constant (maybe it already exists ?).

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 3]
```diff
@@ -427,24 +426,31 @@ private NodeIterator getDigestIterator(NotificationContext context, SessionProvi
...
-    List<AbstractChannel> channels = channelManager.getChannels();
-    for (AbstractChannel channel : channels) {
-      model.setChannelPlugins(channel.getId(), getValues(node, getChannelProperty(channel.getId())));
-    }
-    //
-    model.setLastUpdateTime(node.getParent().getProperty(EXO_LAST_MODIFIED_DATE).getDate());
-    //
-    if (node.hasProperty(EXO_IS_ENABLED)) {
-      model.setEnabled(Boolean.valueOf(node.getProperty(EXO_IS_ENABLED).getString()));
+    if(!node.getParent().hasProperty( EXO_LAST_MODIFIED_DATE)){
```
Comment (thomasdelhomenie) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: We should fetch the parent node once IMO.

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 4]
```diff
@@ -474,9 +480,16 @@ private NodeIterator getDefaultDailyIterator(SessionProvider sProvider, int offs
...
+            if(node.canAddMixin("exo:modify")) {
+              node.addMixin("exo:modify");
+            }
+              node.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
+              node.save();
+          }
+            users.add(UserSetting.getInstance()
                     .setUserId(node.getName())
-                    .setLastUpdateTime(node.getProperty(EXO_LAST_MODIFIED_DATE).getDate()));
+                      .setLastUpdateTime(node.getProperty(EXO_LAST_MODIFIED_DATE).getDate()));
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: Avoid to change file formatting here

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 5]
```diff
@@ -474,9 +480,16 @@ private NodeIterator getDefaultDailyIterator(SessionProvider sProvider, int offs
...
           Node node = iter.nextNode();
-          users.add(UserSetting.getInstance()
+          if(!node.hasProperty( EXO_LAST_MODIFIED_DATE)){
+            if(node.canAddMixin("exo:modify")) {
+              node.addMixin("exo:modify");
+            }
+              node.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
+              node.save();
+          }
+            users.add(UserSetting.getInstance()
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: Avoid to change file formatting here

**Comment by boubaker** [comment]
Please rework the PR without changing the file formatting, space characters ...

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java (hunk 3 above)
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: This comment isn't considered yet ?

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java [hunk 6]
```diff
@@ -245,17 +237,23 @@ private NodeIterator getDailyNodes(Node pluginDayNode, String userId) throws Exc
...
-
-    return message;
+    if(!node.hasProperty( EXO_LAST_MODIFIED_DATE)){
+      if(node.canAddMixin("exo:modify")) {
+        node.addMixin("exo:modify");
+      }
+      node.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
+      node.save();
+    }
+      NotificationInfo message = NotificationInfo.instance()
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java]: Don't change formatting information. Tab = 2 spaces.

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java (hunk 2 above)
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: This comment isn't considered yet ?
use contant: org.exoplatform.services.wcm.core.NodetypeConstant.EXO_MODIFY

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 7]
```diff
@@ -427,24 +426,31 @@ private NodeIterator getDigestIterator(NotificationContext context, SessionProvi
...
-      model.setEnabled(Boolean.valueOf(node.getProperty(EXO_IS_ENABLED).getString()));
+    if(!node.getParent().hasProperty( EXO_LAST_MODIFIED_DATE)){
+      if(node.getParent().canAddMixin("exo:modify")) {
+        node.getParent().addMixin("exo:modify");
+      }
+      node.getParent().setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
+      node.getParent().save();
     }
-    return model;
+      UserSetting model = UserSetting.getInstance();
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: Use the same code formatting please

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 8]
```diff
@@ -474,6 +480,13 @@ private NodeIterator getDefaultDailyIterator(SessionProvider sProvider, int offs
         NodeIterator iter = getDefaultDailyIterator(sProvider, offset, limit);
         while (iter.hasNext()) {
           Node node = iter.nextNode();
+          if(!node.hasProperty( EXO_LAST_MODIFIED_DATE)){
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: Add "node.canAddMixin("exo:modify")" to this if and delete the next "if" statement

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java (hunk 2 above)
Comment (azayati) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: In that case we should add dependency to ECMS core services module. Is it better to add EXO_MODIFY as a constant into org.exoplatform.commons.notification.impl.AbstractService class ?

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java [hunk 9]
```diff
@@ -245,16 +237,20 @@ private NodeIterator getDailyNodes(Node pluginDayNode, String userId) throws Exc
...
     if(node == null) return null;
+    if(!node.hasProperty(EXO_LAST_MODIFIED_DATE) && node.canAddMixin(EXO_MODIFY)){
+      node.addMixin(EXO_MODIFY);
+      node.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
+      node.save();
+    }
     NotificationInfo message = NotificationInfo.instance()
       .setFrom(node.getProperty(NTF_FROM).getString())
-      .setOrder(Integer.valueOf(node.getProperty(NTF_ORDER).getString()))
+	  .setOrder(Integer.valueOf(node.getProperty(NTF_ORDER).getString()))
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java]: Don't change code format plz

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java [hunk 10]
```diff
@@ -245,16 +237,20 @@ private NodeIterator getDailyNodes(Node pluginDayNode, String userId) throws Exc
...
       .key(node.getProperty(NTF_PROVIDER_TYPE).getString())
       .setOwnerParameter(node.getProperty(NTF_OWNER_PARAMETER).getValues())
       .setSendToDaily(NotificationUtils.valuesToArray(node.getProperty(NTF_SEND_TO_DAILY).getValues()))
       .setSendToWeekly(NotificationUtils.valuesToArray(node.getProperty(NTF_SEND_TO_WEEKLY).getValues()))
       .setLastModifiedDate(node.getProperty(EXO_LAST_MODIFIED_DATE).getDate())
       .setId(node.getName());
-
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java]: Don't change code format plz

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 11]
```diff
@@ -427,8 +426,14 @@ private NodeIterator getDigestIterator(NotificationContext context, SessionProvi
    * @throws Exception
    */
   private UserSetting fillModel(Node node) throws Exception {
+	Node parentNode = node.getParent();
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: Use the same code format plz

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 12]
```diff
@@ -427,8 +426,14 @@ private NodeIterator getDigestIterator(NotificationContext context, SessionProvi
...
+    if(!parentNode.hasProperty(EXO_LAST_MODIFIED_DATE) && parentNode.canAddMixin(EXO_MODIFY)){
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: What if the mixin was already added and not the property ?
The property isn't mandatory so the mixin can be added without adding property.
`
if(!parentNode.hasProperty(EXO_LAST_MODIFIED_DATE) ){
  if(parentNode.isNodetype(EXO_MODIFY)) {
    parentNode.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
    parentNode.save();
  } else if(parentNode.canAddMixin(EXO_MODIFY)) {
    parentNode.addMixin(EXO_MODIFY);
    parentNode.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
    parentNode.save();
  } else {
   LOG.warn("Cannot add mixin to node '{}'.", parentNode.getPath());
  }
}
`

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 13]
```diff
@@ -439,7 +444,7 @@ private UserSetting fillModel(Node node) throws Exception {
       model.setChannelPlugins(channel.getId(), getValues(node, getChannelProperty(channel.getId())));
     }
     //
-    model.setLastUpdateTime(node.getParent().getProperty(EXO_LAST_MODIFIED_DATE).getDate());
+    model.setLastUpdateTime(parentNode.getProperty(EXO_LAST_MODIFIED_DATE).getDate());
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: if(parentNode.hasProperty(EXO_LAST_MODIFIED_DATE) ){
    model.setLastUpdateTime(node.getParent().getProperty(EXO_LAST_MODIFIED_DATE).getDate());
} else {
   model.setLastUpdateTime(Calendar.getInstance());
}

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java [hunk 14]
```diff
@@ -474,6 +479,11 @@ private NodeIterator getDefaultDailyIterator(SessionProvider sProvider, int offs
         NodeIterator iter = getDefaultDailyIterator(sProvider, offset, limit);
         while (iter.hasNext()) {
           Node node = iter.nextNode();
+          if(!node.hasProperty(EXO_LAST_MODIFIED_DATE) && node.canAddMixin(EXO_MODIFY)){
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: Same as previous comment:


if(!node.hasProperty(EXO_LAST_MODIFIED_DATE) ){
if(node.isNodetype(EXO_MODIFY)) {
node.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
node.save();
} else if(node.canAddMixin(EXO_MODIFY)) {
node.addMixin(EXO_MODIFY);
node.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
node.save();
} else {
LOG.warn("Cannot add mixin to node '{}'.", node.getPath());
}
}

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java (hunk 9 above)
Comment (azayati) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java]: Done

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java (hunk 10 above)
Comment (azayati) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java]: Done

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java (hunk 11 above)
Comment (azayati) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: done

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java (hunk 12 above)
Comment (azayati) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: done

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java (hunk 13 above)
Comment (azayati) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: done

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java (hunk 14 above)
Comment (azayati) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/setting/UserSettingServiceImpl.java]: done

**Comment by boubaker** [comment]
ok, to be tested.

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java [hunk 15]
```diff
@@ -245,6 +237,11 @@ private NodeIterator getDailyNodes(Node pluginDayNode, String userId) throws Exc
 
   private NotificationInfo fillModel(Node node) throws Exception {
     if(node == null) return null;
+    if(!node.hasProperty(EXO_LAST_MODIFIED_DATE) && node.canAddMixin(EXO_MODIFY)){
+      node.addMixin(EXO_MODIFY);
+      node.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
+      node.save();
+    }
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java]: Use the same strategy here:

if(!node.hasProperty(EXO_LAST_MODIFIED_DATE)) {
  if(node.isNodeType(EXO_MODIFY)) {
	node.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
	node.save();
  }
  else if(node.canAddMixin(EXO_MODIFY)) {
	node.addMixin(EXO_MODIFY);
	node.setProperty(EXO_LAST_MODIFIED_DATE, Calendar.getInstance());
	node.save();
  }
  else {
	LOG.warn("Cannot add mixin to node '{}'.", node.getPath());
  }
}

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java [hunk 16]
```diff
@@ -254,7 +251,7 @@ private NotificationInfo fillModel(Node node) throws Exception {
       .setSendToWeekly(NotificationUtils.valuesToArray(node.getProperty(NTF_SEND_TO_WEEKLY).getValues()))
       .setLastModifiedDate(node.getProperty(EXO_LAST_MODIFIED_DATE).getDate())
       .setId(node.getName());
-
+
```
Comment (boubaker) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java]: Don't change this line too

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java (hunk 15 above)
Comment (azayati) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java]: done

File: commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java (hunk 16 above)
Comment (azayati) [inline commons-component-common/src/main/java/org/exoplatform/commons/notification/impl/service/storage/NotificationDataStorageImpl.java]: done

**Pull request closed without merging**